swarm/network/simulations/overlay_test.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/simulations"
+)
+
+var (
+	nodeCount = 16
+)
+
+//This test is used to test the overlay simulation.
+//As the simulation is executed via a main, it is easily missed on changes
+//An automated test will prevent that
+//The test just connects to the simulations, starts the network,
+//starts the mocker, gets the number of nodes, and stops it again.
+//It also provides a documentation on the steps needed by frontends
+//to use the simulations
+func TestOverlaySim(t *testing.T) {
+	//start the simulation
+	log.Info("Start simulation backend")
+	go runOverlaySim()
+
+	//the simulation runs usually at localhost:8888
+	localhostAddr := fmt.Sprintf("http://localhost:%d", httpSimPort)
+	//wait for the network to start
+	time.Sleep(1 * time.Second)
+
+	log.Debug("Start simulation network")
+	//start the simulation network (initialization of simulation)
+	resp, err := http.Post(localhostAddr+"/start", "application/json", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("Expected Status Code %d, got %d", http.StatusOK, resp.StatusCode)
+	}
+
+	log.Debug("Start mocker")
+	//start the mocker, needs a node count and an ID
+	formstr := []byte(fmt.Sprintf("node-count=%d&mocker-type=%s", nodeCount, simulations.GetMockerList()[0]))
+
+	resp, err = http.Post(localhostAddr+"/mocker/start", "application/x-www-form-urlencoded", bytes.NewReader(formstr))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		reason, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			t.Fatal(err)
+		}
+		log.Error(string(reason))
+		resp.Body.Close()
+		t.Fatalf("Expected Status Code %d, got %d", http.StatusOK, resp.StatusCode)
+	}
+
+	//the mocker needs some time to start up nodes
+	time.Sleep(2 * time.Second)
```

Comment:
can we get the node up events from the sim instead?

Reply:
Yes we can, and I did, although it's a bit overkill IMHO for this simple test. And in fact, ironically I we have to keep a little `Sleep` later, because stopping the network immediately after may cause the sim to crash due to a running node wanting to connect to a already stopped not from `StopNetwork` :)